File: examples/kotlin/src/test/kotlin/adr/spine/ContextTest.kt
// ── test/spine/context — F4/G15: the third projection, and its bound ──────
// The reasoner's input used to be the one seam with no type, no projection, no
// bound, no capture rule and no test layer. This file is the test layer.

package adr.spine

import adr.app.initialState
import adr.app.Env
import adr.app.projectContextApp
import adr.app.Wiring
import adr.blocks.triage.SET_PRIORITY
import adr.blocks.triage.Ticket
import adr.blocks.triage.TriageBlock
import adr.human
import adr.spine.pure.MAX_CONTEXT_LINES_PER_BLOCK
import adr.spine.pure.MAX_CONTEXT_NOTICES
import adr.spine.pure.Notice
import adr.spine.pure.SpineSlice
import adr.spine.pure.SourceName
import adr.spine.pure.StagedInput
import adr.spine.pure.TicketId
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName
import adr.spine.pure.ContextRenderer
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** The blocks that contribute context lines: triage, escalation, console, analysis, inbox. */
private const val CONTEXT_LINE_BLOCKS = 5

class ContextTest {

    @Test
    fun `projectContext is a pure function of committed state plus this turn's staged input`() {
        val state = initialState(listOf(Ticket(TicketId("4118"), "refund not received")))
        val staged = listOf(
            StagedInput.Perceived(SourceName("inbox"), "customer says the refund never arrived"),
        )

        val context = projectContextApp(state, staged)

        assertEquals(staged, context.staged)
        assertEquals(0, context.artifactLineCount)
        assertTrue(context.lines.any { it.contains("ticket 4118") })
        // Calling it again on the same state gives the same value — it is a projection,
        // not an accumulator. Nothing was appended anywhere.
        assertEquals(context, projectContextApp(state, staged))
    }

    @Test
    fun `G15 - the context is BOUNDED, so it does not grow with session length`() {
        val tickets = (1..500).map { Ticket(TicketId("T$it"), "body $it") }
        val noisy = initialState(tickets).let { s ->
            s.copy(
                spine = SpineSlice(
                    run = s.spine.run,
                    notices = (1..200).map { Notice.Rejected(Timestamp(it.toLong()), ToolName("t$it"), "r$it") },
                ),
            )
        }

        val context = projectContextApp(noisy, emptyList())

        assertEquals(MAX_CONTEXT_LINES_PER_BLOCK, TriageBlock().contextLines(noisy.triage).size)
        assertEquals(MAX_CONTEXT_NOTICES, context.notices.size)
        assertTrue(context.lines.size <= CONTEXT_LINE_BLOCKS * MAX_CONTEXT_LINES_PER_BLOCK)
        // The artifact contributes a COUNT, never content — a long session cannot inflate it.
        assertEquals(0, context.artifactLineCount)
        assertTrue(
            ContextRenderer().render(context).lines().size <=
                CONTEXT_LINE_BLOCKS * MAX_CONTEXT_LINES_PER_BLOCK + MAX_CONTEXT_NOTICES + 2,
        )
    }

    @Test
    fun `the rendered digest and the prompt version ride the committed record (14_7)`() {
        val app = Wiring().wireApp(Env())
        app.human(SET_PRIORITY, "ticket" to "4118", "level" to "High")

        val fixture = app.bus.records().single().context
        assertEquals("triage-prompt@1", fixture.promptVersion)
        // The digest is the state as it was BEFORE the step — the input the model saw.
        assertEquals(ContextRenderer().render(projectContextApp(app.initial, emptyList())), fixture.digest)
    }
}

File: examples/typescript/test/spine/context.test.ts
// ── G15 — the reasoner's input is a named, typed, BOUNDED projection ───────
// The shipped reference had one sentence for this seam: no type, no projection,
// no bound, no capture rule, no invariant, no test layer. Below is the test
// layer.

import { describe, expect, it } from "vitest";
import { projectContext } from "../../src/app/assemble";
import { initialState } from "../../src/app/contract";
import {
  MAX_CONTEXT_LINES_PER_BLOCK,
  MAX_CONTEXT_NOTICES,
  render,
} from "../../src/spine/pure/context";
import { rejected } from "../../src/spine/pure/notice";
import { perceived } from "../../src/spine/pure/staged";
import { contextDivergence } from "../../src/spine/replay/replay";
import { harness } from "../harness";

describe("projectContext — the THIRD pure projection (G15)", () => {
  it("is a projection of committed State plus the ONE staged input", () => {
    const state = initialState({ tickets: [{ id: "4118", body: "refund not received" }] });
    const staged = [perceived("inbox", "customer wrote in", "inbox-1")];
    const context = projectContext(state, staged);

    expect(context.staged).toEqual(staged);
    expect(context.artifactLineCount).toBe(0);
    expect(context.lines).toContain("ticket 4118 [Normal]: refund not received");
    expect(context.lines).toContain("ticket 4118: open, may be escalated");
    // calling it twice on the same input is the same value — it accumulates nothing
    expect(projectContext(state, staged)).toEqual(context);
  });

  it("is O(1) in timeline length: 500 tickets and 200 notices stay within the caps", () => {
    const tickets = Array.from({ length: 500 }, (_, i) => ({ id: `t${i}`, body: `body ${i}` }));
    const base = initialState({ tickets });
    const state = {
      ...base,
      spine: {
        ...base.spine,
        notices: Array.from({ length: 200 }, (_, i) => rejected(i, "setPriority", `reason ${i}`)),
      },
    };

    const context = projectContext(state, []);
    // four blocks, each capped
    expect(context.lines.length).toBeLessThanOrEqual(4 * MAX_CONTEXT_LINES_PER_BLOCK);
    expect(context.notices.length).toBe(MAX_CONTEXT_NOTICES);
    // the artifact enters by COUNT, never by content
    expect(context.artifactLineCount).toBe(0);
    expect(render(context).length).toBeLessThan(4000);
  });

  it("the committed digest is re-derivable from committed State — the fixture IS a check", () => {
    const h = harness();
    h.app.boundary.onStepFinish({
      by: "Agent",
      staged: [perceived("inbox", "urgent", "inbox-2")],
      actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
    });
    h.app.boundary.onStepFinish({
      by: "Agent",
      staged: [],
      actions: [{ tool: "recordFinding", input: { text: "noted" } }],
    });

    expect(contextDivergence(h.app.initial, h.app.bus.records(), h.app.dispatchers)).toEqual([]);
  });

  it("a change to what the model saw fails the golden trace without re-running the model", () => {
    const h = harness();
    h.app.boundary.onStepFinish({
      by: "Agent",
      staged: [],
      actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
    });

    const tampered = h.app.bus.records().map((r) => ({
      ...r,
      context: { ...r.context, digest: `${r.context.digest} (and one more thing)` },
    }));
    expect(contextDivergence(h.app.initial, tampered, h.app.dispatchers)).toEqual([
      "step 0: context digest diverged",
    ]);
  });
});
